### PR TITLE
Feat: enable owner references

### LINKED ISSUE #498: Add ownerReference to managed entities
If I delete a database, resources created by it (services, statefulsets, etc.) should be deleted automatically. This can be easily implemented by adding the Postgres resource as an `ownerReference` (with `controller: true`). Kubernetes will then automatically delete the resources once the master database resource is gone.

### PR DESCRIPTION
**Pull request by FxKu**
This a follow up to #2199 with some greater changes. The most important part of owner references is the cascaded removal of child resources which bypasses the delete protection offered by our operator.

Quick reminder about delete protection: this checks pre-defined annotations before calling delete code but cannot block the actual kubectl delete call itself. This you could only achieve with a K8s admission controller which, for example, does the same annotation checks.

We use the latter at Zalando, but some folks out there might rely on the protection by the operator. Therefore, we should not enable owner references by default but make it configurable.

#2199 also lacked code to update existing resources once you start using owner references or vice versa remove them. Like with the recently added annotation inheritance, this produced many code extensions. I’ve also found some oversights from #2657 which I’ve fixed along the way. All owner references syncs are done with Update API calls, which requires extending the RBAC.

A new e2e test is added which tests syncing owner references in both directions and deleting the acid-test-cluster cascadingly which was previously done in the multi namespace test. Some unit test were extended, too.

closes #2199
fixes #498

### CODE REVIEW

File: manifests/operatorconfiguration.crd.yaml [hunk 1]
```diff
@@ -209,9 +209,9 @@ spec:
                   enable_init_containers:
                     type: boolean
                     default: true
-                  enable_secrets_deletion:
+                  enable_owner_refereces:
```
Comment (FxKu) [inline manifests/operatorconfiguration.crd.yaml:212]:
```suggestion
                  enable_owner_references:
```

**Comment by fm2022aa** [comment]
great

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -519,9 +524,13 @@ func (c *Cluster) compareStatefulSetWith(statefulSet *appsv1.StatefulSet) *compa
 				reasons = append(reasons, fmt.Sprintf("new statefulset's name for volume %d does not match the current one", i))
 				continue
 			}
+			if !reflect.DeepEqual(c.Statefulset.Spec.VolumeClaimTemplates[i].OwnerReferences, statefulSet.Spec.VolumeClaimTemplates[i].OwnerReferences) {
```
Comment (hughcapet) [inline pkg/cluster/cluster.go:527]: OwnerReferences on PVC?

**Review by hughcapet — COMMENTED** [review]
Don't we need OwnerReferences on logical backup jobs?

File: pkg/cluster/sync.go [hunk 3]
```diff
@@ -957,9 +970,15 @@ func (c *Cluster) updateSecret(
 		userMap[userKey] = pwdUser
 	}
 
+	if !reflect.DeepEqual(secret.ObjectMeta.OwnerReferences, generatedSecret.ObjectMeta.OwnerReferences) {
+		updateSecret = true
+		updateSecretMsg = fmt.Sprintf("secret %s owner references do not match the current ones", secretName)
```
[the remark this answers is not in the document]
Comment (FxKu) [reply pkg/cluster/sync.go:975]: The other cases are mutual exclusive to one another. This line here is the only case that might happen additionally. But given that enabling owner references is probably a one time operation, I don't think it's worth to add more logic.

File: pkg/cluster/sync.go [hunk 4]
```diff
@@ -232,21 +231,35 @@ func (c *Cluster) syncService(role PostgresRole) error {
...
+		err            error
 	)
 	c.setProcessName("syncing %s endpoint", role)
 
 	if ep, err = c.KubeClient.Endpoints(c.Namespace).Get(context.TODO(), c.endpointName(role), metav1.GetOptions{}); err == nil {
 		desiredEp := c.generateEndpoint(role, ep.Subsets)
-		if changed, _ := c.compareAnnotations(ep.Annotations, desiredEp.Annotations); changed {
-			patchData, err := metaAnnotationsPatch(desiredEp.Annotations)
+		if !reflect.DeepEqual(ep.ObjectMeta.OwnerReferences, desiredEp.ObjectMeta.OwnerReferences) {
+			updateEndpoint = true
```
[the remark this answers is not in the document]
Comment (FxKu) [reply pkg/cluster/sync.go:243]: agree. I added a comment to make it clearer

**Review by hughcapet — APPROVED** [review]

**Comment by FxKu** [comment]
:+1:

**Comment by hughcapet** [comment]
👍

**Comment by FxKu** [comment]
:+1:

**Comment by barthy1** [comment]
Thank you for this great work!
When this feature is going to be released?

**Comment by holgerson97** [comment]
> Thank you for this great work! When this feature is going to be released?

@barthy1 it's already in 1.13.0 and I am testing it.

**Pull request merged**